### PR TITLE
CLDSRV-215 S3.DeleteObjects should clean up data

### PR DESCRIPTION
**Pull request by nicolas2bert**
Deleting a version should delete its data on the backend side.

### CODE REVIEW

**Comment by nicolas2bert** [comment]
ping

File: lib/api/multiObjectDelete.js [hunk 1]
```diff
@@ -233,6 +233,12 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
                         log.debug('trying to delete locked object');
                         return callback(objectLockedError);
                     }
+                    if (versionId && objMD.location &&
+                        Array.isArray(objMD.location) && objMD.location[0]) {
```
Comment (francoisferrand) [inline lib/api/multiObjectDelete.js:237]: nit: may be easier to read with a function "wrapping" the `objMD.location && Array.isArray(objMD.location) && objMD.location[0]` expression behind a meaningful name (like `isRemoteObject` or whatever this condition means)

File: lib/api/multiObjectDelete.js [hunk 2]
```diff
@@ -233,6 +233,12 @@ function getObjMetadataAndDelete(authInfo, canonicalID, request,
...
+                        // we need this information for data deletes to AWS
+                        // eslint-disable-next-line no-param-reassign
+                        objMD.location[0].deleteVersion = true;
```
Comment (francoisferrand) [inline lib/api/multiObjectDelete.js:240]: nit: probably for later, but it seems strange to store this flag only in the first location, it hints that the flag should not be in the 'location' array but separate from it, somehow... → maybe some refactoring to think of in the future.

**Review by francoisferrand — APPROVED** [review]

**Review by naren-rajendran — APPROVED** [review]

File: lib/api/multiObjectDelete.js (hunk 1 above)
Comment (miniscruff) [reply lib/api/multiObjectDelete.js:237]: Small addition to that, undefined is not an array so the third check handles the second one already.

File: tests/functional/aws-node-sdk/test/multipleBackend/delete/deleteAwsVersioning.js [hunk 3]
```diff
@@ -557,3 +583,83 @@ describeSkipIfNotMultiple('AWS backend delete object w. versioning: ' +
...
+    'using object location constraint', function testSuite() {
+    this.timeout(120000);
+    withV4(sigCfg => {
+        let bucketUtil;
+        let s3;
+        beforeEach(() => {
+            process.stdout.write('Creating bucket\n');
+            bucketUtil = new BucketUtility('default', sigCfg);
+            s3 = bucketUtil.s3;
+            return s3.createBucket({ Bucket: bucket }).promise()
```
Comment (miniscruff) [inline tests/functional/aws-node-sdk/test/multipleBackend/delete/deleteAwsVersioning.js:598]: you could use async await here btw, quite minor but you probably wouldn't need to catch the error just to throw it back. If you let the test runner catch the error the call stack is already available.

**Review by miniscruff — APPROVED** [review]

**Comment by nicolas2bert** [comment]
@bert-e approve

**Pull request merged**